File: state.py
import logging
import subprocess
from datetime import datetime
from threading import Lock

log = logging.getLogger(__name__)

_lock = Lock()
_render_lock = Lock()

_cancel_flag: bool = False
_current_render_process: "subprocess.Popen | None" = None

_status: dict = {
    "state": "idle",
    "operation": "",
    "current": 0,
    "total": 0,
    "name": "",
    "out_name": "",
    "size": "",
    "error": "",
    "started_at": "",
    "started_at_ts": 0.0,
    "frame_current": 0,
    "frame_total": 0,
    "frame_pct": 0,
    "eta_seconds": 0,
    "hw_nvdec": False,
    "hw_nvenc": False,
    "hw_trt": False,
    "face_count": 0,
    "plate_count": 0,
    "logs": [],
}


def _log(msg: str) -> None:
    ts = datetime.now().strftime("%H:%M:%S")
    entry = f"[{ts}] {msg}"
    log.info(msg)
    with _lock:
        _status["logs"].append(entry)
        if len(_status["logs"]) > 200:
            _status["logs"] = _status["logs"][-200:]


def _set(**kwargs) -> None:
    with _lock:
        _status.update(kwargs)


def _format_eta(seconds: int) -> str:
    if seconds <= 0:
        return "–"
    if seconds < 60:
        return f"{seconds} Sek"
    m, s = divmod(seconds, 60)
    return f"{m} Min {s} Sek" if s else f"{m} Min"


def request_cancel() -> None:
    global _cancel_flag
    with _lock:
        _cancel_flag = True


def is_cancel_requested() -> bool:
    with _lock:
        return _cancel_flag


def check_and_clear_cancel() -> bool:
    global _cancel_flag
    with _lock:
        val = _cancel_flag
        if val:
            _cancel_flag = False
    return val


def clear_cancel() -> None:
    global _cancel_flag
    with _lock:
        _cancel_flag = False


def set_render_process(proc: "subprocess.Popen | None") -> None:
    global _current_render_process
    with _render_lock:
        _current_render_process = proc


def get_render_process() -> "subprocess.Popen | None":
    with _render_lock:
        return _current_render_process

File: src/state.rs
use std::collections::VecDeque;
use std::sync::{Arc, Mutex};

#[derive(Debug, Clone, serde::Serialize)]
pub struct AppStatus {
    #[serde(skip)]
    pub preview_jpeg: Vec<u8>,
    pub state: String,
    pub current: u32,
    pub total: u32,
    pub name: String,
    pub error: String,
    pub frame_current: u64,
    pub frame_total: u64,
    pub frame_pct: u8,
    pub eta_seconds: u64,
    pub face_count: u64,
    pub plate_count: u64,
    pub hw_nvdec: bool,
    pub hw_nvenc: bool,
    pub hw_trt: bool,
    pub sub_state: String,
    pub out_name: String,
    pub started_at: String,
    pub started_at_ts: f64,
    pub log: VecDeque<String>,
}

impl Default for AppStatus {
    fn default() -> Self {
        Self {
            state: "idle".into(),
            current: 0,
            total: 0,
            name: String::new(),
            error: String::new(),
            frame_current: 0,
            frame_total: 0,
            frame_pct: 0,
            eta_seconds: 0,
            face_count: 0,
            plate_count: 0,
            hw_nvdec: false,
            hw_nvenc: false,
            hw_trt: false,
            sub_state: String::new(),
            out_name: String::new(),
            started_at: String::new(),
            started_at_ts: 0.0,
            log: VecDeque::with_capacity(200),
            preview_jpeg: Vec::new(),
        }
    }
}

pub type SharedState = Arc<Mutex<AppStatus>>;

pub fn new_state() -> SharedState {
    Arc::new(Mutex::new(AppStatus::default()))
}

pub fn log(state: &SharedState, msg: &str) {
    let ts = chrono::Local::now().format("%H:%M:%S").to_string();
    let line = format!("[{ts}] {msg}");
    tracing::info!("{}", msg);
    let mut s = state.lock().unwrap();
    s.log.push_back(line);
    if s.log.len() > 200 {
        s.log.pop_front();
    }
}

pub fn set_idle(state: &SharedState) {
    let mut s = state.lock().unwrap();
    s.state = "idle".into();
    s.name = String::new();
    s.frame_current = 0;
    s.frame_total = 0;
    s.frame_pct = 0;
    s.eta_seconds = 0;
}
